[RFC] Introduce Checkpointable for DCP (#127540)

iamzainhuda opened the pull request:
Summary:
# Introduce Checkpointable interface for DCP to support arbitrary tensor subclasses for checkpointing

**Authors:**
* zainhuda

## **Summary**
This diff adds a CheckpointableTensor interface to allow for future compatibility for any tensor subclass with DCP in a clean and maintainable way.

## **Motivation**
For TorchRec sharding migration from ShardedTensor to DTensor, we create a tensor subclass that is stored by DTensor to support TorchRec's sharding schemes (ex, empty shards, multiple shards on a rank).

## **Proposed Implementation**
View the CheckpointableTensor interface implementation, in which, we introduce the minimal set of methods needed to be compatible with DCP. These methods are expected to implemented by any tensor subclasses and as such are then checkpointable by DCP.

## **Drawbacks**
No drawbacks, it extends functionality in a clean and maintainable way.

## **Alternatives**
Alternative design was creating paths for checking for certain attributes in tensor subclasses which can get messy and hard to maintain/understand why it was there in the first place.


Test Plan:
Sandcastle





cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k LucasLLC

Differential Revision: D57970603

Pulled By: iamzainhuda




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @LucasLLC

iamzainhuda commented:
Copying comments from earlier PR #127628 127540 (had to close because of export issues):

From Lucas:
> One piece of feedback I have is to change CheckpointableTensor to reflect a generic class for which includes non-tensor objects (e.g. just Checkpointable might be reasonable). Unless you think this approach only works for Tensors? The refactor would also avoid confusing users into thinking this is a proper Tensor subclass

Me:
> Yeah this is great, I'll rename it to Checkpointable instead. Approach should work for tensors and non tensors alike.

iamzainhuda commented:
Copying comments from earlier PR https://github.com/pytorch/pytorch/pull/127628 127540 (had to close because of export issues):

From Lucas:

Overall, this looks like a sustainable pattern to me. I think we could even argue for making DTensor and ShardedTensor use the CheckpointableTensor as a mix-in. I'd like to solicit some more feedback before giving it a final stamp

iamzainhuda commented on torch/distributed/checkpoint/planner.py:
```diff
@@ -426,3 +427,31 @@ def commit_tensor(self, read_item: ReadItem, tensor: torch.Tensor) -> None:
         pass
+
+
+class Checkpointable:
```
Initial comment from Lucas:
> I've been leaning against using abstract classes these days. It makes it harder to maintain both consistency and backwards compatibility when we add new methods, but maybe that's just me.

> Does anyone else have this opinion?

Zain:
> I agree, latest revision removes this. And there was actually some compatibility issues with abc.ABC and torch.Tensor subclasses, which removing helps resolve as well.

wanchaol commented on torch/distributed/checkpoint/planner.py:
```diff
@@ -426,3 +427,31 @@ def commit_tensor(self, read_item: ReadItem, tensor: torch.Tensor) -> None:
+        raise NotImplementedError("Checkpointable._create_chunk_list is not implemented")
+
+    @abc.abstractmethod
+    def find_tensor_shard(self, tensor: torch.Tensor, index: MetadataIndex) -> torch.Tensor:
```
can we rename this as "_get_tensor_shard"? so that it's not a public facing API?

wanchaol commented on torch/distributed/checkpoint/planner.py:
```diff
@@ -426,3 +427,31 @@ def commit_tensor(self, read_item: ReadItem, tensor: torch.Tensor) -> None:
         pass
+
+
+class Checkpointable:
```
I wonder if we should make this public in the first place, or we want to keep it private first and public it later? @LucasLLC

LucasLLC replied:
I'm totally not against a private first approach, which could de-risk this impl.

iamzainhuda replied:
Makes sense to me, made private

iamzainhuda replied:
Agreed, made private

wz337 reviewed (approved):
Great! Thanks for introducing this!

iamzainhuda committed:
[RFC] Introduce Checkpointable for DCP (#127628)

Summary:

# Introduce Checkpointable interface for DCP to support arbitrary tensor subclasses for checkpointing

**Authors:**
* zainhuda

## **Summary**
This diff adds a CheckpointableTensor interface to allow for future compatibility for any tensor subclass with DCP in a clean and maintainable way.

## **Motivation**
For TorchRec sharding migration from ShardedTensor to DTensor, we create a tensor subclass that is stored by DTensor to support TorchRec's sharding schemes (ex, empty shards, multiple shards on a rank).

## **Proposed Implementation**
View the CheckpointableTensor interface implementation, in which, we introduce the minimal set of methods needed to be compatible with DCP. These methods are expected to implemented by any tensor subclasses and as such are then checkpointable by DCP.

## **Drawbacks**
No drawbacks, it extends functionality in a clean and maintainable way.

## **Alternatives**
Alternative design was creating paths for checking for certain attributes in tensor subclasses which can get messy and hard to maintain/understand why it was there in the first place.


Test Plan:
Sandcastle





cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k LucasLLC

Reviewed By: wz337

Differential Revision: D57970603

Pulled By: iamzainhuda

fegin commented on torch/distributed/checkpoint/planner.py:
```diff
@@ -426,3 +426,39 @@ def commit_tensor(self, read_item: ReadItem, tensor: torch.Tensor) -> None:
+        """
+        Return a list of WriteItems based on object's contents.
+        """
+        raise NotImplementedError(
```
nit, since this is an `abstractmethod`, ... is enough instead of `NotImplementedError`.

fegin commented on torch/distributed/checkpoint/planner.py:
```diff
@@ -426,3 +426,39 @@ def commit_tensor(self, read_item: ReadItem, tensor: torch.Tensor) -> None:
+        """
+        Return a list of `ChunkStorageMetadata` based on object's contents.
+        """
+        raise NotImplementedError(
```
ditto

fegin commented on torch/distributed/checkpoint/planner.py:
```diff
@@ -426,3 +426,39 @@ def commit_tensor(self, read_item: ReadItem, tensor: torch.Tensor) -> None:
+        """
+        Return a 'torch.Tensor' shard based on 'MetadataIndex'.
+        """
+        raise NotImplementedError(
```
ditto

fegin reviewed (approved):
some nits, LGTM

iamzainhuda commented:
@pytorchmergebot merge

iamzainhuda commented:
> ## Merge failed
> **Reason**: 1 mandatory check(s) failed. The first few are:
> 
> * [pull / linux-focal-cuda12.1-py3.10-gcc9-sm86 / test (default, 2, 5, linux.g5.4xlarge.nvidia.gpu)](https://github.com/pytorch/pytorch/actions/runs/9354969198/job/25750723065)
> 
> Dig deeper by [viewing the failures on hud](https://hud.pytorch.org/pytorch/pytorch/commit/52b495e404ef23d612c3e6cbfbb576d59f8dc318)
> 
> Details for Dev Infra team

Test fails with:
Running test_foreach 1/1 ... [2024-06-03 19:50:55.859463]
Executing ['/opt/conda/envs/py_3.10/bin/python', '-bb', 'test_foreach.py', '-m', 'not serial', '--shard-id=1', '--num-shards=1', '-v', '-vv', '-rfEX', '-p', 'no:xdist', '--use-pytest', '-x', '--reruns=2', '--import-slow-tests', '--import-disabled-tests'] ... [2024-06-03 19:50:55.859894]

PRINTING LOG FILE of test_foreach 1/1 (test/test-reports/test_foreach_1.1_181a6067961d1387_.log)
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/test_foreach.py", line 16, in <module>
    from torch.testing._internal.common_cuda import TEST_MULTIGPU
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_cuda.py", line 8, in <module>
    from torch.testing._internal.common_utils import LazyVal, TEST_NUMBA, TEST_WITH_ROCM, TEST_CUDA, IS_WINDOWS
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 1301, in <module>
    gb_available = torch.cuda.mem_get_info()[1] / 2 ** 30
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/cuda/memory.py", line 684, in mem_get_info
    return torch.cuda.cudart().cudaMemGetInfo(device)
RuntimeError: CUDA error: out of memory
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

Does not look related to this PR

iamzainhuda commented:
@pytorchbot merge -f

iamzainhuda commented:
@pytorchbot merge -f "failing CI not related to diff"
